services/src/main/java/com/broadleafcommerce/subscriptionoperation/service/modification/AbstractModifySubscriptionHandler.java

```diff
@@ -185,4 +205,94 @@ protected abstract String[] getRequiredPermissions(ModifySubscriptionRequest req
+     * 
+     * @param subscription the current subscription
+     * @param contextInfo the context info
+     * @return the generated PriceCartRequest
+     */
+    protected PriceCartRequest buildPriceCartRequest(@lombok.NonNull Subscription subscription,
+            @Nullable ContextInfo contextInfo) {
+        PriceCartRequest priceCartRequest = typeFactory.get(PriceCartRequest.class);
+        priceCartRequest.setCurrency(subscription.getCurrency());
+        return priceCartRequest;
+    }
+
+    protected void hydrateRequest(@NonNull ModifySubscriptionRequest request,
+            @Nullable ContextInfo contextInfo) {
+        setExistingAndNewProductsOnModifySubscriptionRequest(request, contextInfo);
+    }
+
+    protected void setExistingAndNewProductsOnModifySubscriptionRequest(
+            @NonNull ModifySubscriptionRequest request,
+            @Nullable ContextInfo contextInfo) {}
+
+    protected void setExistingAndNewProductsOnModifySubscriptionRequest(
+            @NonNull ModifySubscriptionRequest request,
+            @NonNull CatalogProvider<Product> catalogProvider,
+            @Nullable ContextInfo contextInfo) {
+        Pair<Product, Product> newAndRootItems =
+                readNewAndRootItems(request, catalogProvider, contextInfo);
+
+        request.setProductToChangeTo(newAndRootItems.getLeft());
+        request.setRootProduct(newAndRootItems.getRight());
+    }
+
+    protected Pair<Product, Product> readNewAndRootItems(
+            @NonNull ModifySubscriptionRequest request,
+            @NonNull CatalogProvider<Product> catalogProvider,
+            @Nullable ContextInfo contextInfo) {
+        Subscription subscription = request.getSubscription().getSubscription();
+
+        if (isBroadleafProduct(request.getNewItemRefType())
+                && isBroadleafProduct(subscription.getRootItemRefType())) {
+
+            String rootItemRef = subscription.getRootItemRef();
+            String newItemRef = request.getNewItemRef();
+
+            List<Product> products = catalogProvider.readProductsByIds(
+                    List.of(rootItemRef, newItemRef), null, contextInfo)
+                    .getContent();
+
+            Product newItemProduct = products.stream()
+                    .filter(product -> Objects.equals(product.getId(), newItemRef))
+                    .findFirst().orElseThrow(EntityMissingException::new);
+
+            Product rootProduct = products.stream()
+                    .filter(product -> Objects.equals(product.getId(), rootItemRef))
+                    .findFirst().orElseThrow(EntityMissingException::new);
+
+            return Pair.of(newItemProduct, rootProduct);
+        } else {
+            return readNewAndRootItemsByCustomReferences(request, contextInfo);
+        }
+    }
+
+    protected Pair<Product, Product> readNewAndRootItemsByCustomReferences(
+            @NonNull ModifySubscriptionRequest request,
+            @Nullable ContextInfo contextInfo) {
```

Comment:
I would have assumed they'd still want the `CatalogProvider` or else how will the fetch the products?
```suggestion
    protected <T extends Product> List<T> readNewAndRootItemsByCustomReferences(
            @NonNull ModifySubscriptionRequest request,
            @NonNull CatalogProvider<T> catalogProvider,
            @Nullable ContextInfo contextInfo) {
```

Reply:
My comment is moot with the changes I made in this PR https://github.com/BroadleafCommerce/SubscriptionOperationServices/pull/20.